File: packages/python/src/pharos_discovery/security/blocklist.py
"""Blocklist — maintain a persistent set of blocked MCP server IDs."""

from __future__ import annotations

import json
import threading
from pathlib import Path
from typing import ClassVar


class Blocklist:
    """Thread-safe in-memory blocklist with JSON file persistence.

    Stores ``{server_id: reason}`` pairs. Use :meth:`add`, :meth:`remove`,
    :meth:`is_blocked` for membership management and :meth:`save`/:meth:`load`
    for persistence.
    """

    def __init__(self) -> None:
        self._entries: dict[str, str] = {}
        self._lock = threading.Lock()

    # ------------------------------------------------------------------ API
    def add(self, server_id: str, reason: str) -> None:
        """Block *server_id* for *reason*."""
        with self._lock:
            self._entries[server_id] = reason

    def remove(self, server_id: str) -> None:
        """Unblock *server_id* (no-op if not present)."""
        with self._lock:
            self._entries.pop(server_id, None)

    def is_blocked(self, server_id: str) -> bool:
        with self._lock:
            return server_id in self._entries

    def list(self) -> dict[str, str]:
        """Return a shallow copy of the blocklist."""
        with self._lock:
            return dict(self._entries)

    # ------------------------------------------------------------ persistence
    def save(self, path: str) -> None:
        """Write the blocklist to *path* as JSON."""
        with self._lock:
            data = dict(self._entries)
        Path(path).write_text(json.dumps(data, indent=2, sort_keys=True))

    @classmethod
    def load(cls, path: str) -> "Blocklist":
        """Load a blocklist from *path* (returns empty instance if missing)."""
        inst = cls()
        p = Path(path)
        if p.exists():
            data = json.loads(p.read_text())
            with inst._lock:
                inst._entries = {str(k): str(v) for k, v in data.items()}
        return inst

    # ----------------------------------------------------------- dunders
    def __len__(self) -> int:
        with self._lock:
            return len(self._entries)

    def __contains__(self, server_id: object) -> bool:
        return self.is_blocked(str(server_id))

    def __repr__(self) -> str:  # pragma: no cover - trivial
        with self._lock:
            return f"Blocklist(entries={len(self._entries)})"

File: packages/typescript/src/security/blocklist.ts
/**
 * Blocklist — maintain a persistent set of blocked MCP server IDs.
 *
 * In-memory `Map` with JSON-file persistence (uses dynamic `node:fs` import
 * so the module stays importable in pure-browser contexts).
 */

const DEFAULT_REASON = "blocked";

/** Dynamic fs loader — avoids static `node:` imports for browser compat. */
async function loadFs(): Promise<{
  writeFile(path: string, data: string, encoding: string): Promise<void>;
  readFile(path: string, encoding: string): Promise<string>;
}> {
  const mod = "node:fs/promises";
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return await (import(mod) as Promise<any>);
}

export class Blocklist {
  private entries = new Map<string, string>();

  /** Block `serverId` for `reason`. */
  add(serverId: string, reason: string = DEFAULT_REASON): void {
    this.entries.set(serverId, reason);
  }

  /** Unblock `serverId` (no-op if absent). */
  remove(serverId: string): void {
    this.entries.delete(serverId);
  }

  /** Whether `serverId` is currently blocked. */
  isBlocked(serverId: string): boolean {
    return this.entries.has(serverId);
  }

  /** Shallow copy of the blocklist as a plain object. */
  list(): Record<string, string> {
    const out: Record<string, string> = {};
    for (const [k, v] of this.entries) out[k] = v;
    return out;
  }

  /** Number of blocked servers. */
  get size(): number {
    return this.entries.size;
  }

  /** Persist the blocklist to `path` as JSON (Node only). */
  async save(path: string): Promise<void> {
    const data: Record<string, string> = this.list();
    const fs = await loadFs();
    await fs.writeFile(path, JSON.stringify(data, null, 2), "utf-8");
  }

  /** Load a blocklist from `path` (returns empty instance if missing). */
  static async load(path: string): Promise<Blocklist> {
    const inst = new Blocklist();
    try {
      const fs = await loadFs();
      const raw = await fs.readFile(path, "utf-8");
      const data = JSON.parse(raw) as Record<string, string>;
      for (const [k, v] of Object.entries(data)) inst.entries.set(k, v);
    } catch {
      // missing or unreadable — return empty
    }
    return inst;
  }

  /** In-memory load from a JSON string (useful for tests / browser). */
  static fromJSON(json: string): Blocklist {
    const inst = new Blocklist();
    try {
      const data = JSON.parse(json) as Record<string, string>;
      for (const [k, v] of Object.entries(data)) inst.entries.set(k, v);
    } catch {
      // ignore parse errors
    }
    return inst;
  }

  /** Serialise to a JSON string. */
  toJSON(): string {
    return JSON.stringify(this.list(), null, 2);
  }
}
